WpfTeslaCamViewer/MainWindow.xaml.cs

```
@@ -1,255 +1,274 @@
-            playerInfoTimer.Elapsed += async (_, _) =>
+            await Application.Current.Dispatcher.BeginInvoke(() =>
             {
-                if (player == null) return;
+                lbl_DebugInfo.Content = info;
+                slider_progress.Value = position;
+            });
+        };
+        playerInfoTimer.Start();
+    }
 
-                var info = player.GetDebugInfo();
-                var position = player.GetCurrentVideoPosition();
+    public ObservableCollection<string> FolderNames { get; set; } = new();
+    public HashSet<string> FileNames { get; set; } = new();
 
-                await Application.Current.Dispatcher.BeginInvoke(() => { 
-                    lbl_DebugInfo.Content = info;
-                    slider_progress.Value = position;
-                });
-            };
-            playerInfoTimer.Start();
-        }
+    private void Window_Loaded(object sender, RoutedEventArgs e)
+    {
+        playerFront = new MediaPlayer(libVlc);
+        playerLeft = new MediaPlayer(libVlc);
+        playerRight = new MediaPlayer(libVlc);
+        playerBack = new MediaPlayer(libVlc);
 
-        private void btn_SlowDown_Click(object sender, RoutedEventArgs e)
-        {
-            if (playbackSpeed > 0.1f)
-            {
-                playbackSpeed -= 0.1f;
-                SetPlaybackRate();
-            }
-        }
+        videoViewFront.MediaPlayer = playerFront;
+        videoViewLeftRepeater.MediaPlayer = playerLeft;
+        videoViewRightRepeater.MediaPlayer = playerRight;
+        videoViewRear.MediaPlayer = playerBack;
 
-        private void btn_SpeedUp_Click(object sender, RoutedEventArgs e)
-        {
-            playbackSpeed += 0.1f;
-            SetPlaybackRate();
-        }
+        player = new TeslaCamPlayer(libVlc, playerFront, playerLeft, playerRight, playerBack,
+            () => Application.Current.Dispatcher.BeginInvoke(GoToNextClip).GetAwaiter().GetResult());
```

I changed this to use the non-async version due to this last parameter being an `Action` which returns `void`, and `async void`s will swallow all exceptions 😬